libbeat/testing/integration/integration.go

```diff
@@ -0,0 +1,336 @@
+	// expectations will be printed before the test starts.
+	//
+	// Use it only if you have a manageable amount of expectations that
+	// would be readable in the output.
+	PrintExpectationsBeforeStart bool
+
+	// PrintLinesOnFail defines how many lines of the Beat output
+	// the test should print in case of failure (default 0).
+	//
+	// It uses `PrintOutput`, see its documentation for details.
+	PrintLinesOnFail int
+}
+
+// BeatTestOptions describes all options to run the test
+type BeatTestOptions = RunBeatOptions
+
+// NewBeatTest creates a new integration test for a Beat.
+func NewBeatTest(t *testing.T, opts BeatTestOptions) BeatTest {
+	test := &beatTest{
+		t:    t,
+		opts: opts,
+	}
+
+	return test
+}
+
+type beatTest struct {
+	t                *testing.T
+	opts             BeatTestOptions
+	reportOpts       ReportOptions
+	expectations     []OutputWatcher
+	expectedExitCode *int
+	beat             *RunningBeat
+	mtx              sync.Mutex
+}
+
+// Start implements the BeatTest interface.
+func (b *beatTest) Start(ctx context.Context) BeatTest {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+	if b.beat != nil {
+		b.t.Fatal("test cannot be startd multiple times")
+		return b
+	}
+	watcher := NewOverallWatcher(b.expectations)
+	b.t.Logf("running %s integration test...", b.opts.Beatname)
+	if b.reportOpts.PrintExpectationsBeforeStart {
+		b.printExpectations()
+	}
+	b.beat = RunBeat(ctx, b.t, b.opts, watcher)
+
+	return b
+}
+
+// Wait implements the BeatTest interface.
+func (b *beatTest) Wait() {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	if b.beat == nil {
+		b.t.Fatal("test must start first before calling wait on it")
+		return
+	}
+
+	err := b.beat.Wait()
+	exitErr := &exec.ExitError{}
+	if !errors.As(err, &exitErr) {
+		b.t.Fatalf("unexpected error when stopping %s: %s", b.opts.Beatname, err)
+		return
+	}
+
+	exitCode := 0
+	if err != nil {
+		exitCode = exitErr.ExitCode()
+	}
+	b.t.Logf("%s stopped, exit code %d", b.opts.Beatname, exitCode)
+
+	if b.expectedExitCode != nil && exitCode != *b.expectedExitCode {
+		b.t.Cleanup(func() {
+			b.t.Logf("expected exit code %d, actual %d", b.expectedExitCode, exitCode)
+		})
+
+		b.t.Fail()
+	}
+
+	if b.beat.watcher != nil {
+		b.t.Cleanup(func() {
+			b.t.Logf("\n\nExpectations are not met:\n\n%s\n\n", b.beat.watcher.String())
+			if b.reportOpts.PrintLinesOnFail != 0 {
+				b.PrintOutput(b.reportOpts.PrintLinesOnFail)
+			}
+		})
+		b.t.Fail()
+	}
+}
+
+// ExpectOutput implements the BeatTest interface.
+func (b *beatTest) ExpectOutput(lines ...string) BeatTest {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+
+	if b.beat != nil {
+		b.t.Fatal(expectErrMsg)
+		return b
+	}
+
+	if len(lines) == 0 {
+		return b
+	}
+
+	if len(lines) == 1 {
+		l := escapeJSONCharacters(lines[0])
+		b.expectations = append(b.expectations, NewStringWatcher(l))
+		return b
+	}
+
+	watchers := make([]OutputWatcher, 0, len(lines))
+	for _, l := range lines {
+		escaped := escapeJSONCharacters(l)
+		b.t.Log(escaped)
```

Comment:
I'm just wondering if we need this log or if it was some debug leftover

Reply:
Good catch, it's a left over from my debugging.